### ISSUE TITLE
[Records] get_record for specific symbol

### ISSUE THREAD

**Issue by qrebjock** [JuliaManifolds/Manopt.jl#70]
Hello, it's me again!

From what I understand in the documentation, the following code

```julia
using LinearAlgebra
using Manopt, Manifolds
using Random

m = 5
n = 5
k = 2

function F(M, p)
    return norm(p.S)^2 / 2 - dot(p.S, diag(p.U' * A * p.Vt'))
end

function gradF(M, p::SVDMPoint)
    dS = diagm(p.S)
    return UMVTVector(
        -(I - p.U * p.U') * A * p.Vt',
        dS - p.U' * A * p.Vt',
        -p.U' * A * (I - p.Vt' * p.Vt)
    )
end

A = randn(m, n);
M = FixedRankMatrices(m, n, k)
X0 = random_point(M);

record = gradient_descent(
    M,
    F,
    gradF,
    X0,
    retraction_method=PolarRetraction(),
    record=[:Iteration, :Cost],
    return_options=true
)

get_record(record, :Cost)
```
should return an array containing the sequence of costs.
But it returns an array of tuples containing all the records (:Iteration and :Cost).
Is it the intended behavior?

**Comment by qrebjock** [comment]
Recording gradient norms is pretty common to evaluate the convergence rate of an algorithm.
Right now it is only possible for the gradient descent scheme I think. With trust regions there is no such option.
What do you think about moving the lines
```julia
@doc raw"""
    RecordGradientNorm <: RecordAction

record the current gradient norm value.
"""
mutable struct RecordGradientNorm <: RecordAction
    recorded_values::Array{Float64,1}
    RecordGradientNorm() = new(Array{Float64,1}())
end
function (r::RecordGradientNorm)(p::P, o::O, i::Int) where {P<:Problem,O<:Options}
    return record_or_reset!(r, norm(p.M, o.x, o.gradient), i)
end
```
to `record_options.jl`?
I guess you haven't done it for a good reason but maybe there is a way to fix it.

**Comment by qrebjock** [comment]
Recording gradient norms is pretty common to evaluate the convergence rate of an algorithm.
Right now it is only possible for the gradient descent scheme I think. With trust regions there is no such option.
What do you think about moving the lines
```julia
@doc raw"""
    RecordGradientNorm <: RecordAction

record the current gradient norm value.
"""
mutable struct RecordGradientNorm <: RecordAction
    recorded_values::Array{Float64,1}
    RecordGradientNorm() = new(Array{Float64,1}())
end
function (r::RecordGradientNorm)(p::P, o::O, i::Int) where {P<:Problem,O<:Options}
    return record_or_reset!(r, norm(p.M, o.x, o.gradient), i)
end
```
to `record_options.jl`?
I guess you haven't done it for a good reason but maybe there is a way to fix it.

**Comment by kellertuer** [comment]
Concerning your first question – whether that is the intended behaviour – yes it is. But I had to check my design, still, too.

The record is intended to work as follows: you can basically record whatever you want, using a `RecordGroup`, which stores a Tuple of values per iteration. This is usually done in the `:All` field of the record dictionary. This field name is not related to field names in the options.
The only other existing field (maybe existing) is `:Stop`, which records the stopping criterion's reason to stop. The reason for this split is, that of course this second one occurs at other times (only once in the end) and is surely of different type than the general recording `:All`.

Since `:Cost` is just a shorthand for `RecordCost()`(and yields a field in the `:All` dictionary antry) nothing really maps to the entry in the recorded tuple to be adressible to refer back to the cost. You just have to know which position in the tuple is which value. For that we would have to switch from said tuple to an own (sub-)dictionary (array of dictionaries?).

We can surely discuss to improve this, but I hope this makes the current design more clear. In hindsight `:All` should be called `:IterationValues` or something. So this symbol is more like the type of recording, not the actual fields or values.

**Comment by kellertuer** [comment]
> What do you think about moving the lines [...] to `record_options.jl`?

We can surely do that. This `RecordAction` is currently already usable with all `Options` that have a `o.gradient` field. Though with 0.3.0 it would fit better do use `get_gradient`. I missed that one when refactoring the `gradient`/`gradient!` stuff. Changing this, would make the record action even more usable for other options.

And yes, to increase visibility, it would be good to move it – but as I said it can already be used with all options. One does just not directly find it in the docs.

**Comment by qrebjock** [comment]
Thanks for the explanations. Yes the current design is more clear now and it makes sense.

Right now calling `get_record(record, :Cost)` and `get_record(record)` return the same array of tuples (in the particular example I showed). This may be a bit confusing for users. Maybe it would be nice to have `get_record(record)` returning an array with all the records (which is already the case), and `get_record(record, :Cost)` returning an array with only the costs?

**Comment by kellertuer** [comment]
The second part is not really possible, since the symbol is used for something different (`:All` should be renamed to `:Iterates`, since it refers more to when the recorded values are stored). So `get_record` can give you all recorded values for the iterations (the `:All` which should be named `:Iterations`) or a recorded `:Stop`ping reason, but it actually should error on `:Cost`.

Let me illustrate : If you would call a solver with
`record = [:Iteration, :Cost, :Gradient]`
let's say you store this in `rec1`, and
`record = [:Iteration, :Gradient, :Cost]`

Then in both cases currently it would record a Tuple per iteration but there is (during recording or in the array of tuples afterwards) no Information anywhere which entry of the tuple corresponds to what semantic value (which you would like to have with `: Cost`), since the tuple per iteration can not know that.

One would have to change the tuple we currently record per iteration into a dictionary (with symbols as keys). This would mean to redesign the Record Actions, since that would mean to redesign the `RecordGroup`, which has to additional store this semantic symbol for each of its inner record actions.
We could do that for sure. It would just mean some work. Then having the `RecordGroup` – let's call it `recG` one could o a `get_record(recG,:Cost)` sure.

Would that work? I think it would be a little work, but not that complicated (redesign the inner struct a little and introduce a `get_record` for the group not only for the `RecordOptions`).

**Comment by qrebjock** [comment]
Ok I see. I think that from an user point of view having such a feature would be nice. So that it's very easy to access recorded series. I don't know if it makes the code less efficient.

I will maybe try to implement this based on your suggestions in the next days, depending on how much time I can allocate for this.

**Comment by kellertuer** [comment]
Oh, great if you could get that started, and no worries, if it takes longer, I am happy that you want to start this.
Let's summarise

* Given `RecordOptions ro`, calling `get_records(ro)`should default to `get_record(ro,:Iterations)`, where `:Iterations` is the new name of `:All`
* `get_record(o,:Iterations)` should return the `RecordGroup rg` that contains the array of dictionaries.
* `get_record(rg, :Cost)` should return an array of cost values.
 
I See one challenge here, namely how to couple a symbol and an action within the `RecordGroup`, i.e. how to couple :Cost to a specific Record Action, such that that action is stored at that symbol. This affects both the `RecordGroup` constructor as well as the long form in the high level interface. My idea

* Currently you can do `record = [:Cost, MyRecordAction()]` in the high level. What about `record = [:Cost, :symbol => MyRecordAction()]`?

Finally a technical proposal: What about you join the project here as a member and do the branch directly here? Then I could also push to that (I can't push to your fork), and its easier to develop together.

Again – thanks for the proposal and ideas and don't worry about the time. If it takes longer – it takes longer. I am more interested in having a feature than in having it fast :)

**Comment by kellertuer** [comment]
I have an idea, one could also stick to the tuples in recording (for performance) and store in a dictionary `:Symbol -> Index`. I hope to find maybe a day over easter to work on this.

**Comment by qrebjock** [comment]
Hi kellertuer,

Sorry for the very late answer. I was quite busy recently.

Thank you very much for taking care of this! It looks very convenient to use, yes.
The idea you have looks like a good compromise.

**Comment by kellertuer** [comment]
No problem, I know there is sometimes things that are keeping us busy. Great that you like it. I thin this way the tuples still keep the recording somewhat efficient, while accessing recorded values got a little nicer. Thanks for the idea.
